core/state/statedb.go

```diff
@@ -973,24 +982,83 @@ func (db *DB) AddReward(snapshot *stk.ValidatorWrapper, reward *big.Int, shareLo
 		rewardPool.Sub(rewardPool, commissionInt)
 	}
 
+	// no sanity check for length of delegations between curValidator and snapshot
+	// if a delegation happens, len(curValidator) > len(snapshot)
+	// if it doesn't happen, curValidator == snapshot
+	// if there are only removals, curValidator < snapshot
+
 	// Payout each delegator's reward pro-rata
 	totalRewardForDelegators := big.NewInt(0).Set(rewardPool)
-	for i := range snapshot.Delegations {
-		delegation := snapshot.Delegations[i]
-		percentage, ok := shareLookup[delegation.DelegatorAddress]
+	if !nilDelegationsRemoved {
+		for i := range snapshot.Delegations {
+			delegation := snapshot.Delegations[i]
+			percentage, ok := shareLookup[delegation.DelegatorAddress]
 
-		if !ok {
-			return errors.Wrapf(err, "missing delegation shares for reward distribution")
-		}
+			if !ok {
+				return errors.Wrapf(err, "missing delegation shares for reward distribution")
+			}
 
-		rewardInt := percentage.MulInt(totalRewardForDelegators).RoundInt()
-		curDelegation := curValidator.Delegations[i]
-		curDelegation.Reward.Add(curDelegation.Reward, rewardInt)
-		rewardPool.Sub(rewardPool, rewardInt)
+			rewardInt := percentage.MulInt(totalRewardForDelegators).RoundInt()
+			curDelegation := curValidator.Delegations[i]
+			// should we check that curDelegation.DelegatorAddress == delegation.DelegatorAddress ?
+			// wasn't there originally so I leave it for now
+			curDelegation.Reward.Add(curDelegation.Reward, rewardInt)
+			rewardPool.Sub(rewardPool, rewardInt)
+		}
+	} else {
+		// iterate simply over snapshot delegations
+		// those added later to curValidator are new delegations which do not receive rewards immediately
+		// those removed from curValidator are stale delegations which do not receive rewards anyway
+		offset := 0
+		for i := 0; i < len(snapshot.Delegations); i++ {
+			delegationFromSnapshot := snapshot.Delegations[i]
+			percentage, ok := shareLookup[delegationFromSnapshot.DelegatorAddress]
+			if !ok {
+				return errors.Wrapf(err, "missing delegation shares for reward distribution")
+			}
+			if percentage.IsZero() { // stale delegation
```

Comment:
So the assumption here is that all the 0 delegation are removed, right?

Reply:
Yes